core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
           versions().subList(versions().size() - versionHistorySizeToKeep, versions().size());
       List<ViewHistoryEntry> history =
           history().subList(history().size() - versionHistorySizeToKeep, history().size());
-      return ImmutableViewMetadata.builder().from(this).versions(versions).history(history).build();
+      return ViewMetadata.buildFrom(this)
+          .discardChanges()
+          .versions(versions)
+          .history(history)
+          .build();
     }
 
     return this;
   }
+
+  static Builder builder() {
+    return new Builder();
+  }
+
+  static Builder buildFrom(ViewMetadata base) {
+    return new Builder(base);
+  }
+
+  class Builder {
+    private int formatVersion = 1;
+    private String location;
+    private Integer currentSchemaId;
+    private List<Schema> schemas = Lists.newArrayList();
+    private int currentVersionId = 1;
+    private List<ViewVersion> versions = Lists.newArrayList();
+    private List<ViewHistoryEntry> history = Lists.newArrayList();
+    private Map<String, String> properties = Maps.newHashMap();
+    private List<MetadataUpdate> changes = Lists.newArrayList();
+
+    private Builder() {}
+
+    private Builder(ViewMetadata base) {
+      this.formatVersion = base.formatVersion();
+      this.location = base.location();
+      this.currentSchemaId = base.currentSchemaId();
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.currentVersionId = base.currentVersionId();
+      this.versions = Lists.newArrayList(base.versions());
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder formatVersion(int newFormatVersion) {
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder location(String newLocation) {
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder currentVersionId(int versionId) {
+      this.currentVersionId = versionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(versionId));
+      return this;
+    }
+
+    public Builder currentSchemaId(Integer schemaId) {
```

Comment:
Wait, do we need a current schema ID? Seems like we should always use the schema ID from the current version, so the current version ID functions as both.

Reply:
In that case maybe we shouldn't even send the `current-schema-id` in https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/view/ViewMetadataParser.java#L68?
That would mean that `ViewMetadata#currentSchemaId()` would not be settable and would become:
```
default Integer currentSchemaId() {
    return currentVersion().schemaId();
  }
```

Reply:
we discussed this offline and I will open a separate PR to remove `current-schema-id` from sending/reading it in the `ViewMetadataParser` and we won't allow setting the current schema id through the builder, because the schema id will be derived from the `ViewVersion`